src/main/scala/nl.knaw.dans.easy.solr4files/ApplicationWiring.scala

```diff
@@ -58,36 +57,71 @@ class ApplicationWiring(configuration: Configuration)
       feedbackMessage <- updateFiles(bag, ddm, filesXML)
       _ <- commit()
     } yield feedbackMessage
+  }.recoverWith {
+    case t: SolrCommitException => Failure(t)
+    case t =>
+      commit()
+      Failure(t)
   }
 
-  def delete(bagId: String): Try[FeedBackMessage] = for {
-    _ <- deleteBag(bagId)
-    _ <- commit()
-  } yield s"Deleted file documents for bag $bagId"
+  def delete(bagId: String): Try[FeedBackMessage] = {
+    for {
+      _ <- deleteBag(bagId)
+      _ <- commit()
+    } yield s"Deleted file documents for bag $bagId"
+  }.recoverWith {
+    case t: SolrCommitException => Failure(t)
+    case t =>
+      commit()
+      Failure(t)
+  }
 
-  private def updateStores(storeNames: Seq[String]): Try[Seq[FeedBackMessage]] = {
-    storeNames
+  private def updateStores(storeNames: Seq[String]): Try[FeedBackMessage] = {
+    val (thrown, results) = storeNames
+      .toStream
       .map(initSingleStore)
-      .collectResults
-      .recoverWith { case t: CompositeException =>
-        throw WrappedCompositeException(s"Tried to update ${ storeNames.size } stores,", t)
-      }
+      .takeUntilFailure
+
+    val stats = s"Updated ${ storeNames.size } stores"
+
+    if (thrown.isEmpty) Success(stats)
+    else if (results.isEmpty) Failure(thrown.get)
+    else {
+      results.foreach(x => logger.info(x))
+      Failure(AnotherFailedException(stats, thrown.get))
+    }
   }
 
-  private def updateBags(storeName: String, bagIds: Seq[String]): Try[Seq[FeedBackMessage]] = {
-    bagIds
+  private def updateBags(storeName: String, bagIds: Seq[String]): Try[FeedBackMessage] = {
+    val (thrown, results) = bagIds
+      .toStream
       .map(uuid => update(storeName, uuid))
-      .collectResults
-      .recoverWith { case t: CompositeException =>
-        throw WrappedCompositeException(s"Tried to update ${ bagIds.size } bags,", t)
-      }
+      .takeUntilFailure
+
+    val stats = s"Updated ${ bagIds.size } bags for $storeName"
+
+    if (thrown.isEmpty) Success(stats)
+    else if (results.isEmpty) Failure(thrown.get)
+    else Failure(AnotherFailedException(stats, thrown.get))
   }
 
-  private def updateFiles(bag: Bag, ddm: DDM, filesXML: Elem) = {
-    (filesXML \ "file")
+  private def updateFiles(bag: Bag, ddm: DDM, filesXML: Elem): Try[FeedBackMessage] = {
+    val (thrown, results) = (filesXML \ "file")
```

Comment:
same here?

Reply:
Would like it, but in this third case the partitioned results make it more complicated.

Reply:
We'll look at it on Monday to see what we can do. Honestly I didn't really look at this one as it looked the same as the one before...

Reply:
```scala
val (thrown, results) = (filesXML \ "file").<etc.>

// note these lazy keywords!
lazy val (withContent, justMetadata) = results.partition(_.isInstanceOf[SubmittedWithContent])
lazy val stats = s"bag ${ bag.bagId }: updated ${ withContent.size } files with content, ${ justMetadata.size } without content"

(thrown, results) match {
  case (Some(e), Seq()) => Failure(e)
  case (Some(e), rs) =>
    for (c <- withContent)
      logger.info(c.solrId)
    Failure(AnotherFailedException(stats, e))
  case (None, _) => Success(stats)
}
```